sympy/physics/mechanics/joint.py

```diff
@@ -0,0 +1,356 @@
+        >>> J2 = PinJoint('J2', rod, plate, parent_joint_pos=lC, parent_axis=rod.frame.z, child_axis=plate.frame.z)
+        >>> J1.kdes()
+        [omega_J1(t) - Derivative(theta_J1(t), t)]
+        >>> J2.kdes()
+        [omega_J2(t) - Derivative(theta_J2(t), t)]
+        >>> rod.masscenter.vel(C.frame)
+        (h/4 - lB/2)*omega_J1(t)*rod_frame.x
+        >>> plate.masscenter.vel(C.frame)
+        ((h/4 - lB/2)*omega_J1(t) + (h/4 + lB/2)*omega_J1(t))*rod_frame.x
+        >>> rod.frame.ang_vel_in(C.frame)
+        omega_J1(t)*C_frame.y
+        >>> plate.frame.ang_vel_in(C.frame)
+        omega_J2(t)*rod_frame.z + omega_J1(t)*C_frame.y
+        >>> rod.frame.dcm(C.frame)
+        Matrix([
+        [cos(theta_J1(t)), 0, -sin(theta_J1(t))],
+        [               0, 1,                 0],
+        [sin(theta_J1(t)), 0,  cos(theta_J1(t))]])
+        >>> plate.frame.dcm(C.frame)
+        Matrix([
+        [ cos(theta_J1(t))*cos(theta_J2(t)), sin(theta_J2(t)), -sin(theta_J1(t))*cos(theta_J2(t))],
+        [-sin(theta_J2(t))*cos(theta_J1(t)), cos(theta_J2(t)),  sin(theta_J1(t))*sin(theta_J2(t))],
+        [                  sin(theta_J1(t)),                0,                   cos(theta_J1(t))]])
+
+    Notes
+    ======
+
+    All dynamicsymbols used in `Joints` are `real ` by default. It is adviced to use
+    `real=True` assumption when passing dynamicsymbols to `Joints` else it may give
+    incorrect result in some cases.
+
+    """
+
+    def __init__(self, name, parent, child, coordinates=None, speeds=None, parent_joint_pos=None,
+        child_joint_pos=None, parent_axis=None, child_axis=None):
+
+        super().__init__(name, parent, child, coordinates, speeds, parent_joint_pos, child_joint_pos,
+            parent_axis, child_axis)
+
+    def _generate_coordinates(self, coordinate):
+        coordinates = []
+        if coordinate is None:
+            theta = dynamicsymbols('theta' + '_' + self._name, real=True)
+            coordinate = theta
+        coordinates.append(coordinate)
+        return coordinates
+
+    def _generate_speeds(self, speed):
+        speeds = []
+        if speed is None:
+            omega = dynamicsymbols('omega' + '_' + self._name, real=True)
+            speed = omega
+        speeds.append(speed)
+        return speeds
+
+    def _orient_frames(self):
+        self._child.frame.orient_axis(self._parent.frame, self._parent_axis, 0)
+        angle, axis = self._align_axis(self._parent_axis, self._child_axis)
+        if axis != Vector(0):
+            I = ReferenceFrame('I')
+            I.orient_axis(self._child.frame, self._child_axis, 0)
+            I.orient_axis(self._parent.frame, axis, angle)
+            self._child.frame.orient_axis(I, self._parent_axis, self._coordinates[0])
+        else:
+            self._child.frame.orient_axis(self._parent.frame, self._parent_axis, self._coordinates[0])
```

Comment:
Writing out the steps so I understand what you are doing.

For the case where axis == 0, the child and parent joint axes are rotated integer multiples of pi wrt to each other. So you:

1. Align the child frame to the parent frame.
2. Calculate the angle between the two joint rotation axes and the axis of rotation (when the two frames are aligned)
3. Orient the child frame relative to the parent frame about the parent joint axis through the joint angle.

Step 3 overwrites step 1 and does not seem to take into account if the child joint axis and the parent joint axis have opposite directions.

For the case when axis != 0 you do:

1. Align the child frame to the parent frame.
2. Calculate the angle between the two joint rotation axes and the axis of rotation (when the two frames are aligned).
3. Create a new intermediate reference frame and align it with the child reference frame. (now child, parent, and intermediate are all aligned the same).
4. Orient the intermediate frame relative to the parent through the alignment angle about the alignment axis.
5. Orient the child frame relative to the intermediate frame about the parent joint axis through the joint angle. 

Step 4 creates a kinematic orientation loop that is invalid because you already have the child/parent orientation in step 1 set. I hope everything is being properly overwritten so the incorrect orientations do not remain. This is more elegant than what I was doing today, as it only requires 1 intermediate frame. I can only hope that's what Sahil and I did years back! :)

Side note: why are you using `._child` and `.parent` here? If these are public attributes (should be), then you should be using `.child` and `.parent` here.